controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,405 @@
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        this.state.updateAndGet(x -> {
+            if (Objects.equals(x.getVersion(), newState.getVersion())) {
+                result.complete(null);
+                return new Data<>(newState.getData(), newState.getVersion() + 1);
+            } else {
+                result.completeExceptionally(new WriteConflictException("state"));
+                return x;
+            }
+        });
+
+        return result;
     }
 
     @Override
-    public synchronized CompletableFuture<StreamConfiguration> getConfiguration() {
-        return CompletableFuture.completedFuture(this.configuration);
+    CompletableFuture<Data<Integer>> getStateData() {
+        if (this.state.get() == null) {
+            return FutureHelpers.failedFuture(new DataNotFoundException(getName()));
+        }
+
+        return CompletableFuture.completedFuture(state.get());
     }
 
     @Override
-    public CompletableFuture<Boolean> updateState(State state) {
-        this.state = state;
-        return CompletableFuture.completedFuture(true);
+    CompletableFuture<Void> createSegmentTable(Create create) {
+        final int numSegments = create.getConfiguration().getScalingPolicy().getMinNumSegments();
+        final double keyRangeChunk = 1.0 / numSegments;
+
+        final int startingSegmentNumber = 0;
+        final List<AbstractMap.SimpleEntry<Double, Double>> newRanges = IntStream.range(0, numSegments)
+                .boxed()
+                .map(x -> new AbstractMap.SimpleEntry<>(x * keyRangeChunk, (x + 1) * keyRangeChunk))
+                .collect(Collectors.toList());
+
+        segmentTable.set(new Data<>(TableHelper.updateSegmentTable(startingSegmentNumber,
+                new byte[0],
+                newRanges,
+                create.getCreationTime()), 0));
+
+        return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public CompletableFuture<State> getState() {
-        return CompletableFuture.completedFuture(state);
+    CompletableFuture<Segment> getSegmentRow(int number) {
+        return getSegmentTable()
+                .thenApply(x -> TableHelper.getSegment(number, x.getData()));
     }
 
     @Override
-    public synchronized CompletableFuture<Segment> getSegment(int number) {
-        return CompletableFuture.completedFuture(segments.get(number));
+    CompletableFuture<Data<Integer>> getSegmentTable() {
+        if (this.segmentTable.get() == null) {
+            return FutureHelpers.failedFuture(new DataNotFoundException(getName()));
+        }
+
+        return CompletableFuture.completedFuture(segmentTable.get());
     }
 
     @Override
-    public synchronized CompletableFuture<Integer> getSegmentCount() {
-        return CompletableFuture.completedFuture(segments.size());
+    CompletableFuture<Data<Integer>> getSegmentTableFromStore() {
+        return getSegmentTable();
```

Comment:
fixed it, here and everywhere else (getSegmentTable returns copy of the data in the store.)